### PR TITLE
Added regex to NFA conversion

### PR DESCRIPTION
**Pull request by jjaworska**

### CODE REVIEW

File: src/main/kotlin/cacophony/automata/NFAConstructor.kt [hunk 1]
```diff
@@ -0,0 +1,135 @@
+package cacophony.automata
+
+import kotlin.collections.emptyMap
+import cacophony.utils.AlgebraicRegex
+
+fun buildNFAFromRegex(regex: AlgebraicRegex): NFA<Int> = buildSimpleNFAFromRegex(regex)
+
+// implementation relies on the fact that the initial state will always have number 0
```
Comment (Giselus) [inline src/main/kotlin/cacophony/automata/NFAConstructor.kt:8]: Please start comment with uppercase letter and end it with dot to stay consistent with other PRs.

File: src/test/kotlin/cacophony/automata/NFAConstructorTest.kt [hunk 2]
```diff
@@ -0,0 +1,114 @@
...
+            var states = epsClosure(setOf(nfa.getStartingState()))
+            for (symbol in str) {
+                states = step(states, symbol)
+            }
+            return (nfa.getAcceptingState() in states)
+        }
+    }
+
+    @Test
+    fun `Simple atomic regex test`() {
```
Comment (Giselus) [inline src/test/kotlin/cacophony/automata/NFAConstructorTest.kt:41]: We also start test function names from lowercase letter.

File: src/main/kotlin/cacophony/automata/NFAConstructor.kt
Comment (Giselus) [inline src/main/kotlin/cacophony/automata/NFAConstructor.kt:1]: Looks good, but I'm not sure about this file name. There is already simpleDFA class, and it's in the DFA.kt. It doesn't have to be like that, but usally filename is the same as main interface/class name, and we don't have anything like NFAConstructor. Maybe we can move buildNFAFromRegex to NFA.kt and rename this file to simpleNFA.kt? Just suggestion, probably can be done differently.

**Review by Giselus — CHANGES_REQUESTED** [review]
Looks mainly good, only cosmetic changes. But please, fix the build, as it's currently failing due to linter.

File: src/main/kotlin/cacophony/automata/NFAConstructor.kt [hunk 3]
```diff
@@ -0,0 +1,135 @@
...
+                epsProds.getOrPut(sub.getAcceptingState()) { mutableListOf() }.add(finalState)
+            }
+            SimpleNFA(
+                initState,
+                prods,
+                epsProds,
+                finalState)
+        }
+        is AlgebraicRegex.ConcatenationRegex -> {
+            var prods: Map<Pair<Int, Char>, List<Int>> = mapOf()
```
Comment (Giselus) [inline src/main/kotlin/cacophony/automata/NFAConstructor.kt:43]: Maybe it should be mutableMap?

**Review by Giselus — CHANGES_REQUESTED** [review]

File: src/main/kotlin/cacophony/automata/NFAConstructor.kt
Comment (jjaworska) [reply src/main/kotlin/cacophony/automata/NFAConstructor.kt:1]: Following the convention which was chosen for DFA's, I moved the SimpleNFA class to the NFA.kt, but left the implementation in a separate file, didn't really have a name better idea

**Review by Giselus — APPROVED** [review]

**Pull request merged**